src/scaffold/admission_request.rs

```diff
@@ -0,0 +1,801 @@
+use std::{
+    collections::HashMap,
+    fmt::{self, Display, Formatter},
+    fs::File,
+    future::Future,
+    path::PathBuf,
+    str::FromStr,
+};
+
+use anyhow::{anyhow, Result};
+use directories::ProjectDirs;
+use k8s_openapi::{
+    api::authentication::v1::UserInfo,
+    apimachinery::pkg::{apis::meta::v1::APIResource, runtime::RawExtension},
+};
+use kube::api::DynamicObject;
+use lazy_static::lazy_static;
+use policy_evaluator::admission_request::{
+    AdmissionRequest, GroupVersionKind, GroupVersionResource,
+};
+use policy_evaluator::kube;
+use serde::{Deserialize, Serialize};
+use tracing::{debug, info, warn};
+
+lazy_static! {
+    pub static ref DEFAULT_ROOT: ProjectDirs =
+        ProjectDirs::from("io.kubewarden", "", "kubewarden").unwrap();
+    pub static ref DEFAULT_KWCTL_CACHE: PathBuf = DEFAULT_ROOT.cache_dir().join("kwctl");
+    pub static ref RESOURCE_CATALOG_FILE: PathBuf =
+        DEFAULT_KWCTL_CACHE.join("resource_catalog.json");
+}
+
+const FALLBACK_API_RESOURCE_PLURAL_NAME: &str = "this-is-the-plural-name-of-the-resource-this-information-is-not-used-by-policies-and-requires-a-connection-to-an-api-server-to-be-obtained";
+
+/// Types of AdmissionRequest operation we can scaffold
+pub(crate) enum Operation {
+    Create,
+    Update,
+    Delete,
+}
+
+impl FromStr for Operation {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "CREATE" => Ok(Operation::Create),
+            "UPDATE" => Ok(Operation::Update),
+            "DELETE" => Ok(Operation::Delete),
+            _ => Err(format!("Invalid operation: {}", s)),
+        }
+    }
+}
+
+impl Display for Operation {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            Operation::Create => write!(f, "CREATE"),
+            Operation::Update => write!(f, "UPDATE"),
+            Operation::Delete => write!(f, "DELETE"),
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+enum ApiResourceCatalogRestoredFrom {
+    #[default]
+    Cache,
+    ApiServer,
+    Empty,
+}
+
+/// A catalog of Kubernetes resources. The catalog is built once by querying a Kubernetes API server.
+///
+/// This is required because some information about the resource being scaffolded cannot be
+/// inferred from the object itself. For example: knowning if a resource is namespaced or not, or
+/// the plural name of the resource.
+#[derive(Serialize, Deserialize, Debug)]
+struct ApiResourceCatalog {
+    resources: HashMap<String, APIResource>,
+    #[serde(skip)]
+    restored_from: ApiResourceCatalogRestoredFrom,
+}
+
+impl ApiResourceCatalog {
+    /// Create a new catalog. The function will try to load the catalog from disk, if it fails
+    /// it will build it by querying the Kubernetes API server.
+    /// If querying the API server fails, the function will return an empty catalog.
+    pub async fn new<F, Fut>(resource_catalog_file: PathBuf, build_kubeclient_fn: F) -> Self
```

Comment:
let's add a comment explaining why we need to defer the creation of the client passing a fn here.

Reply:
done